src/runtime_src/core/tools/xbutil2/OO_MemWrite.cpp

```diff
@@ -104,49 +124,149 @@ XBU::verbose("SubCommand option: read mem");
+  uint64_t addr = 0;
   try {
     //-- base address
-    addr = std::stoll(m_baseAddress, nullptr, 0);
+    addr = std::stoull(m_baseAddress, nullptr, 0);
   }
-  catch(const std::invalid_argument&) {
+  catch (const std::invalid_argument&) {
     std::cerr << boost::format("ERROR: '%s' is an invalid argument for '--address'\n") % m_baseAddress;
     throw xrt_core::error(std::errc::operation_canceled);
   }
 
+  // Validate the number of bytes to be written if defined
+  // This does not need to be defined for the --input option path
+  uint64_t size = 0;
   try {
-    //-- size
-    size = std::stoll(m_sizeBytes, nullptr, 0);
+    if (!m_sizeBytes.empty())
+      size = XBUtilities::string_to_base_units(m_sizeBytes, XBUtilities::unit::bytes);
   }
-  catch(const std::invalid_argument&) {
-    std::cerr << boost::format("ERROR: '%s' is an invalid argument for '--size'\n") % m_sizeBytes;
+  catch (const xrt_core::error& e) {
+    std::cerr << boost::format("Value supplied to --size is invalid: %s\n") % e.what();
     throw xrt_core::error(std::errc::operation_canceled);
   }
 
+  if (size == 0) {
+    std::cerr << boost::format("ERROR: Value for --size must be greater than 0\n");
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
+  // Parse the input option path
+  if (!m_inputFile.empty()) {
+    // Verify that the file exists and is not a directory
+    if (!boost::filesystem::exists(m_inputFile) && boost::filesystem::is_regular_file(m_inputFile)) {
+      std::cerr << boost::format("Input file does not exist: '%s'\n") % m_inputFile;
+      throw xrt_core::error(std::errc::operation_canceled);
+    }
+
+    // Open the input file stream after validating the file path and name
+    std::ifstream input_stream(m_inputFile, std::ios::binary);
+    // If count is unspecified, calculate based on the file size and block size
+    uint64_t count = m_count;
+    if (vm["count"].defaulted()) {
+      input_stream.seekg(0, input_stream.end);
+      // tellg returns a signed value as the number of bytes. Validate the return code and
+      // cast it into a useful format
+      auto nonvalidated_length = input_stream.tellg();
+      if (nonvalidated_length < 0)
+        throw std::runtime_error("Failed to get input file length");
+      uint64_t validated_length = static_cast<uint64_t>(nonvalidated_length);
+
+      if (m_sizeBytes.empty()) // update size
+        size = validated_length;
+
+      // Set count such that the entire input stream is written to the device
+      // Add size to the validated length to account for truncation due to division
+      // If the given size is 0 set the block count to 0
+      count = (validated_length + size - 1) / size;
+      input_stream.seekg(0, input_stream.beg);
+    }
+    else
+      count = m_count;
+
+    // Logging information
+    XBU::verbose(boost::str(boost::format("Device: %s") % xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(device))));
```

Comment:
Update the XBUtilities::verbose to support both "std::string" or "boost::format" arguments. 
Doing so will clean up this code nicely :)